app/services/owner_classifier.rb

```diff
@@ -4,7 +4,27 @@ def initialize(id_and_type)
   end
 
   def user_or_usergroup
-    return nil unless @id_and_type =~ /^\d+-(Users|Usergroups)$/
+    validate_input_format!
+    owner = find_owner_by_id
+    raise_owner_not_found_error!(owner)
+    owner
+  end
+
+  private
+
+  def raise_owner_not_found_error!(owner)
+    raise ActiveRecord::RecordNotFound, _("User or Usergroup not found.") unless owner
+  end
+
+  def validate_input_format!
+    raise ArgumentError, _("Invalid input format. Please use the format '${id}-[Users|Usergroups]'.") unless valid_input?
+  end
+
+  def find_owner_by_id
     @id_and_type.include?('Users') ? User.find_by_id(@id_and_type.to_i) : Usergroup.find_by_id(@id_and_type.to_i)
   end
+
+  def valid_input?
+    @id_and_type =~ /^\d+-(Users|Usergroups)$/
```

Comment:
I know you're refactoring, but while you're at it it's better to use `.match?` if you don't use the captured value. In Ruby 3.2+ you also can't use `nil =~` anymore, or on other non-string values. So this is future proofing it.
```suggestion
    @id_and_type.is_a?(String) && @id_and_type.match?(/^\d+-(Users|Usergroups)$/)
```